### ISSUE TITLE
Cant select external/sdcard storage

### ISSUE THREAD

**Issue by kevinclevenger** [flackbash/AudioAnchor#111]
Yes, I have read through #2 & #21. There's no point at which external storage becomes available. Voice, Smart Audiobook, Pulsar, Rocket, FoldPlay, etc all find external storage with no issues.

Android 11, Samsung s10e.

**Comment by HogoDorn** [comment]
Same problem here.
 Samsung Tab S7+, Android 11

**Comment by flackbash** [comment]
I'll have a look at it. Although it might be difficult to fix for me in case it's connected to Android 11.
Thanks for reporting this :)

**Comment by DeadNews** [comment]
Same on Android 11.

**Comment by flackbash** [comment]
This is indeed due to heavy changes about which app can access which parts of the storage introduced in Android 11.
So right now it is not possible to select the sd card in the file picker unless it's formatted as internal storage.
I'll try to fix this.

Note to self: [This](https://stackoverflow.com/questions/63815276/how-to-get-dir-file-list-in-android-11) might help.

**Comment by pjones** [comment]
@flackbash I'm experiencing this issue as well.  How can I help?

**Comment by glimmerwell** [comment]
Same issue here. Thanks for trying to fix it!

Galaxy S20+ 5G
Android 11

**Comment by flackbash** [comment]
@pjones very kind of you. So far, AA relies on getting the path to a directory that contains music. It sounds like this is not possible more with Android 11 (unless one requires heavy permissions). This means extensive structural changes are needed in order to fix this issue. Unfortunately, I don't have a lot of time on my hands at the moment so I haven't gotten around to fixing this yet. What could possibly take some work load off me is some pointers to code snippets that show how similar apps deal with these new Android 11 rules.
Or of course if you're experienced in Android development and have a lot of free time feel free to give it a try and open a PR.
Apart from that I think there's not much you can do right now, but thanks anyway :)

**Comment by meximan** [comment]
SD card selection was working before the last update

**Comment by flackbash** [comment]
@meximan For you selecting SD card was working with AudioAnchor version 2.2.4 and with 2.2.5 not anymore? And you are on Android 11?

**Comment by meximan** [comment]
Its OK I moved on to a better audiobook player thats smoother and more stable: voice.

**Comment by IcedQuinn** [comment]
Can confirm on an S7 Lite running 11 you can only access `/storage/emulated/0/`

From looking around at other applications it looks like as of 11 you're meant to use the system filepicker to select directories which also handles dispensing permissions for the specific folders involved (c.f. https://developer.android.com/guide/topics/providers/document-provider)

**Comment by meximan** [comment]
I guess it depends on the source code for the location of the files or something. Voice (another audiobook app) doesn't have this issue.

**Comment by LeSnake04** [comment]
As of android 11 the old file api got severely limited and is considered depricated.

Apps should use SAF instead.

**Comment by flackbash** [comment]
I agree, although "basically unusable" is a bit of an overstatement in my opinion. As long as you have your audio files in internal storage (or your SD card formatted as internal storage) everything works fine, right?

**Comment by LeSnake04** [comment]
Its pretty common to store music and audio books on the sd card for easy transfer when switching phones or for putting music on ur new phone.

Unless your phone provider decided 0.1 mm of thines is more important than a the years of history on sd cards...

**Comment by hcmh** [comment]
Would it be possible to enable a text input field for a included folders? Because I restored the database and the folders on the SD-Card work just fine, I just cannot add a new one as I cannot add the path. But I would imaging that it might work, if I could just type the path instead.

**Comment by huftis** [comment]
I too have now upgraded to a phone with Android 11 and no longer have access to any files on my SD card from AudioAnchor. :cry: The SD card is simply not shown as option when using the ‘add folder’ feature.

BTW, I also use Simple File Manager ([F-Droid](https://f-droid.org/en/packages/com.simplemobiletools.filemanager.pro/) / [source code](https://github.com/SimpleMobileTools/Simple-File-Manager)). This has a very similar ‘select folder’ feature, where I *can* select folders and files on my SD card. I only had to give a ‘All files’ access to the app to make this work. So it’s certainly possible to request access to all files. I don’t know if this is of any help, but thought I would mention it.

**Comment by flackbash** [comment]
Thanks for the link. I'll give this issue another look once I have more time on my hands (sadly, that might take a while) and check how it is solved in the Simple File Manager app, maybe that will help.

@hcmh Good to know! Although that would be a very dirty hack, but better than nothing of course.

**Comment by glimmerwell** [comment]
I everyone, I hope you are doing well. I am still following this issue. 

If it helps, [Voice Audiobook Player](https://f-droid.org/en/packages/de.ph1b.audiobook/), allows accessing Audiobooks on the SD card. 
Source code: https://github.com/PaulWoitaschek/Voice

I have been using Voice Audiobook Player for now, but it does not have anywhere near the awesome features that Audio Anchor has. Hope you have not given up on your great app!

**Comment by flackbash** [comment]
I have not given up on it, I'll give this issue another try when I have some more time on my hands. Probably beginning of April.

**Comment by flackbash** [comment]
Update: I spent a holiday looking into this recently. However, the approach I followed where I tried to change as little of the general structure of AudioAnchor and its database as possible turned out to be very tedious, would require a ton of changes anyway and on top of that would likely severely impact performance.

The problem is not even mainly to be able to select directories from the external storage (that I solved), but to be able to (quickly) access these and their included files later on (e.g. in the MainActivity of the app, when you select an album in your library i.e. in the AlbumActivity, or when you select a track for playback, i.e. in the PlayActivity, so basically everywhere).

To solve this would probably take me a lot of time and frankly more time than I'm currently willing to put into it. Therefore, I'm afraid there won't be a solution to this issue in the foreseeable future (unless of course someone else takes it upon themselves to fix this :) ).

I can recommend formatting your SD card as internal storage. This should solve the issue and in my opinion comes with several advantages. But I understand that this is not a solution for everyone.

**Comment by sezanzeb** [comment]
> To solve this would probably take me a lot of time and frankly more time than I'm currently willing to put into it

I feel you. Thanks for everything you did so far on the app

**Comment by DearFox** [comment]
Hello, are there any updates regarding external storage support?

**Comment by just-Nob** [comment]
Looks like there is a PM with a working soliton.

**Comment by DearFox** [comment]
It's great to hear we're getting SD card support! My new phone doesn't have the option to install an SD card, but it would definitely make that app even better :3
